src/protocols/unreal2/types.rs

```diff
@@ -0,0 +1,256 @@
+use crate::buffer::Buffer;
+use crate::errors::GDErrorKind::PacketBad;
+use crate::protocols::types::{CommonPlayer, CommonResponse, ExtraRequestSettings, GenericPlayer};
+use crate::protocols::GenericResponse;
+use crate::{GDError, GDResult};
+
+use super::Unreal2StringDecoder;
+
+use std::collections::{HashMap, HashSet};
+
+use byteorder::ByteOrder;
+
+/// Unreal 2 packet types.
+#[derive(Clone, Copy, Debug, Hash, PartialEq, PartialOrd)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+#[repr(u8)]
+pub enum PacketKind {
+    ServerInfo = 0,
+    MutatorsAndRules = 1,
+    Players = 2,
+}
+
+impl TryFrom<u8> for PacketKind {
+    type Error = GDError;
+    fn try_from(value: u8) -> GDResult<Self> {
+        match value {
+            0 => Ok(Self::ServerInfo),
+            1 => Ok(Self::MutatorsAndRules),
+            2 => Ok(Self::Players),
+            _ => Err(PacketBad.context("Unknown packet type")),
+        }
+    }
+}
+
+/// Unreal 2 server info.
+#[derive(Clone, Debug, PartialEq)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+pub struct ServerInfo {
+    pub server_id: u32,
+    pub ip: String,
+    pub game_port: u32,
+    pub query_port: u32,
+    pub name: String,
+    pub map: String,
+    pub game_type: String,
+    pub num_players: u32,
+    pub max_players: u32,
+}
+
+impl ServerInfo {
+    pub fn parse<B: ByteOrder>(buffer: &mut Buffer<B>) -> GDResult<Self> {
+        let server_id = buffer.read()?;
+        let ip = buffer.read_string::<Unreal2StringDecoder>(None)?;
+        let game_port = buffer.read()?;
+        let query_port = buffer.read()?;
+        let name = buffer.read_string::<Unreal2StringDecoder>(None)?;
+        let map = buffer.read_string::<Unreal2StringDecoder>(None)?;
+        let game_type = buffer.read_string::<Unreal2StringDecoder>(None)?;
+        let num_players = buffer.read()?;
+        let max_players = buffer.read()?;
+
+        Ok(ServerInfo {
+            server_id,
+            ip,
+            game_port,
+            query_port,
+            name,
+            map,
+            game_type,
+            num_players,
+            max_players,
+        })
```

Comment:
As there isn't any usage of index skipping or multi-line processing, we could move these variable declarations directly inside the return block to make it a bit more tidier.
```rust
Ok(ServerInfo {
    server_id: buffer.read()?,
    ip: buffer.read_string::<Unreal2StringDecoder>(None)?
  //...
```

Reply:
I wasn't sure if that could cause it to execute out of order, but yeah that would be cleaner.